### PR TITLE
Fixed #31716 -- Fixed detection of console scripts in autoreloader on Windows.

### PR DESCRIPTION
**Pull request by orf**
Ticket: https://code.djangoproject.com/ticket/31716

### CODE REVIEW

File: docs/releases/3.0.8.txt [hunk 1]
```diff
@@ -18,3 +18,5 @@ Bugfixes
 * Reallowed, following a regression in Django 3.0, non-expressions having a
   ``filterable`` attribute to be used as the right-hand side in queryset
   filters (:ticket:`31664`).
+
+* Fixed a crash in ``django-admin runserver`` on Windows (:ticket:`31716`).
```
Comment (carltongibson) [inline docs/releases/3.0.8.txt:22]: I'm not sure if this qualifies for a backport. This wouldn't have worked for many an age no? 🤔

File: django/utils/autoreload.py [hunk 2]
```diff
@@ -207,12 +207,27 @@ def get_child_arguments():
...
+        exe_entrypoint = py_script.with_suffix('.exe')
+        script_entrypoint = py_script.with_name('{}-script.py'.format(py_script.name))
+        if exe_entrypoint.exists():
+            # Should be executed directly, ignoring sys.executable
+            return [exe_entrypoint] + sys.argv[1:]
+        elif script_entrypoint.exists():
+            # Should be executed as usual
+            return args + [script_entrypoint] + sys.argv[1:]
+        else:
+            raise RuntimeError("Script {} does not exist.".format(py_script))
```
Comment (felixxm) [inline django/utils/autoreload.py:230]: `"Script {} does not exist.".format(py_script)` ->
`'Script %s does not exist." % py_script`

File: docs/releases/3.0.8.txt (hunk 1 above)
Comment (felixxm) [inline docs/releases/3.0.8.txt:22]: I'm not sure about backporting, I know that it's a crashing bug but it's not a regression and I'm afraid we can introduce new regressions.

File: django/utils/autoreload.py [hunk 3]
```diff
@@ -207,12 +207,27 @@ def get_child_arguments():
...
+    py_script = Path(sys.argv[0])
 
     args = [sys.executable] + ['-W%s' % o for o in sys.warnoptions]
-    if sys.argv[0] == django.__main__.__file__:
+    if py_script == django_main_path:
         # The server was started with `python -m django runserver`.
         args += ['-m', 'django']
         args += sys.argv[1:]
+    elif not py_script.exists():
+        # sys.argv[0] can not exist for several reasons on Windows.
```
Comment (felixxm) [inline django/utils/autoreload.py:219]: `can not` -> `cannot`, or better `may not `

File: django/utils/autoreload.py (hunk 2 above)
Comment (felixxm) [inline django/utils/autoreload.py:230]: Can we reorganize this a bit?
```python
        exe_entrypoint = py_script.with_suffix('.exe')
        if exe_entrypoint.exists():
            # Should be executed directly, ignoring sys.executable
            return [exe_entrypoint, *sys.argv[1:]]
        script_entrypoint = py_script.with_name('%s-script.py' % py_script.name)
        if script_entrypoint.exists():
            # Should be executed as usual
            return [*args, script_entrypoint, *sys.argv[1:]]
        raise RuntimeError('Script %s does not exist.'% py_script)
```

File: tests/utils_tests/test_autoreload.py [hunk 4]
```diff
@@ -360,11 +409,14 @@ def patch_autoreload(self, argv):
...
 
     def test_manage_py(self):
+        script = Path('manage.py')
+        script.touch()
         argv = ['./manage.py', 'runserver']
         mock_call = self.patch_autoreload(argv)
         autoreload.restart_with_reloader()
         self.assertEqual(mock_call.call_count, 1)
         self.assertEqual(mock_call.call_args[0][0], [self.executable, '-Wall'] + argv)
+        script.unlink()
```
Comment (felixxm) [inline tests/utils_tests/test_autoreload.py:419]: `try: ... finally:` ?

**Comment by carltongibson** [comment]
OK, just confirming, the fix looks good on Windows, and I can't see it broke anything elsewhere... so 👍

Tidy up, remove release notes, squash. 

Thanks @orf 🥇

**Comment by orf** [comment]
Thanks for the review! Happy with not backporting, I only did this because the bug was marked against 3.0 in Trac.

Currently on holiday for a week with no laptop, I will fix it up when I’m back.

**Comment by carltongibson** [comment]
I can do it, don't worry. Enjoy your holiday! 🏖

File: tests/utils_tests/test_autoreload.py (hunk 4 above)
Comment (orf) [reply tests/utils_tests/test_autoreload.py:419]: Or `self.addCleanup(script.unlink)` might be better?

File: tests/utils_tests/test_autoreload.py [hunk 5]
```diff
@@ -153,6 +154,54 @@ def test_path_with_embedded_null_bytes(self):
...
+        with mock.patch("sys.argv", [script_path.with_name("django-admin"), "runserver"]):
+            self.assertEqual(
+                autoreload.get_child_arguments(),
+                [sys.executable, script_path, "runserver"]
+            )
+
+    @mock.patch("sys.argv", ["does-not-exist", "runserver"])
+    @mock.patch("sys.warnoptions", [])
+    def test_raises_runtimeerror(self):
+        with self.assertRaises(RuntimeError, msg="Script does-not-exist does not exist."):
```
Comment (felixxm) [inline tests/utils_tests/test_autoreload.py:201]:
```python
msg = 'Script does-not-exist does not exist.'
with self.assertRaises(RuntimeError, msg):
```

File: tests/utils_tests/test_autoreload.py (hunk 4 above)
Comment (carltongibson) [reply tests/utils_tests/test_autoreload.py]: I went with `addCleanup()` — Thanks for mentioning that @orf.

**Pull request merged**